File: core/src/main/java/me/huanmeng/gui/util/item/ItemBuilder.java
package me.huanmeng.gui.util.item;

import me.huanmeng.gui.adventure.ComponentConvert;
import net.kyori.adventure.text.Component;
import org.bukkit.*;
import org.bukkit.block.banner.Pattern;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.*;
import org.bukkit.map.MapView;
import org.bukkit.potion.PotionEffect;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.util.*;

/**
 * 2023/8/24<br>
 * Gui<br>
 *
 * @author huanmeng_qwq
 */
public class ItemBuilder {

    private ItemStack is;

    public ItemBuilder(ItemStack is) {
        this.is = is.clone();
    }

    public ItemBuilder(ItemStack is, String name) {
        this.is = is;
        setName(name);
    }

    public ItemBuilder(Material m) {
        is = new ItemStack(m);
    }

    public ItemBuilder(Material m, String name) {
        is = new ItemStack(m);
        setName(name);
    }

    public ItemBuilder(Material m, int amount) {
        is = new ItemStack(m, amount);
    }

    public ItemBuilder(Material m, int amount, byte durability) {
        is = new ItemStack(m, amount, durability);
    }

    public ItemBuilder(Material type, int amount, short durability) {
        is = new ItemStack(type, amount, durability);
    }

    public ItemBuilder clone() {
        return new ItemBuilder(is);
    }

    public ItemBuilder setDurability(short dur) {
        is.setDurability(dur);
        return this;
    }

    public ItemBuilder setDurability(int dur) {
        is.setDurability((short) dur);
        return this;
    }

    public ItemBuilder setAmount(int amount) {
        is.setAmount(amount);
        return this;
    }

    public ItemBuilder setName(String str) {
        String name = str;
        ItemMeta im = is.getItemMeta();
        if (name != null && name.isEmpty()) {
            name = "§c";
        }
        if (name == null) {
            im.setDisplayName(null);
        } else {
            im.setDisplayName(ChatColor.translateAlternateColorCodes('&', name));
        }
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder setName(Component component) {
        return setName(ComponentConvert.getDefault().convert(component));
    }

    public ItemBuilder addLore(List<Component> lores) {
        addLore(lores.stream().map(c -> ComponentConvert.getDefault().convert(c)).toArray(String[]::new));
        return this;
    }

    public ItemBuilder setNameUnColor(String name) {
        ItemMeta im = is.getItemMeta();
        if (name != null && name.isEmpty()) {
            name = "§c";
        }
        if (name == null) {
            im.setDisplayName(null);
        } else {
            im.setDisplayName(name);
        }
        is.setItemMeta(im);
        return this;
    }

    public String getName() {
        ItemMeta im = is.getItemMeta();
        return im != null ? im.getDisplayName() : null;
    }

    public ItemBuilder addFlag(ItemFlag... flags) {
        ItemMeta im = is.getItemMeta();
        im.addItemFlags(flags);
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder removeFlag(ItemFlag... flags) {
        ItemMeta im = is.getItemMeta();
        im.removeItemFlags(flags);
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder addUnsafeEnchantment(Enchantment ench, int level) {
        if (is.getType() == Material.ENCHANTED_BOOK) {
            EnchantmentStorageMeta im = (EnchantmentStorageMeta) is.getItemMeta();
            im.addStoredEnchant(ench, level, true);
            is.setItemMeta(im);
            return this;
        } else {
            is.addUnsafeEnchantment(ench, level);
            return this;
        }
    }

    public ItemBuilder addRandomUnsafeEnchantment(int level, Enchantment... ench) {
        is.addUnsafeEnchantment(ench[new Random().nextInt(ench.length)], level);
        return this;
    }

    public ItemBuilder clearLores() {
        ItemMeta im = is.getItemMeta();
        im.setLore(new ArrayList<>());
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder removeEnchantment(Enchantment ench) {
        is.removeEnchantment(ench);
        return this;
    }

    public ItemBuilder removeEnchantments() {
        ItemMeta im = is.getItemMeta();
        im.getEnchants().forEach((enchantment, integer) -> im.removeEnchant(enchantment));
        return this;
    }

    public ItemBuilder setSkullOwner(String owner) {
        try {
            SkullMeta im = (SkullMeta) is.getItemMeta();
            im.setOwner(owner);
            is.setItemMeta(im);
        } catch (ClassCastException ignored) {
        }
        return this;
    }

    public ItemBuilder setSkullOwner(OfflinePlayer player) {
        try {
            SkullMeta im = (SkullMeta) is.getItemMeta();
            Object playerProfile = getPlayerProfileMethod.bindTo(player).invoke();
            setPlayerProfileMethod.bindTo(im).invoke(playerProfile);
            is.setItemMeta(im);
        } catch (Throwable ignored) {
            setSkullOwner(player.getName());
        }
        return this;
    }

    public ItemBuilder setSkullPlayerProfile(Object playerProfile) {
        try {
            SkullMeta im = (SkullMeta) is.getItemMeta();
            setPlayerProfileMethod.bindTo(im).invoke(playerProfile);
            is.setItemMeta(im);
        } catch (Throwable throwable) {
            throw new RuntimeException(throwable);
        }
        return this;
    }

    public ItemBuilder addEnchant(Enchantment ench, int level) {
        if (is.getType() == Material.ENCHANTED_BOOK) {
            EnchantmentStorageMeta im = (EnchantmentStorageMeta) is.getItemMeta();
            im.addStoredEnchant(ench, level, true);
            is.setItemMeta(im);
            return this;
        } else {
            ItemMeta im = is.getItemMeta();
            im.addEnchant(ench, level, true);
            is.setItemMeta(im);
            return this;
        }
    }

    public ItemBuilder addEnchantments(Map<Enchantment, Integer> enchantments) {
        is.addEnchantments(enchantments);
        return this;
    }

    public ItemBuilder setInfinityDurability() {
        is.setDurability((short) 32767);
        return this;
    }

    public ItemBuilder setUnbreakable() {
        ItemMeta im = is.getItemMeta();
        try {
            itemUnbreakableMethod.bindTo(im).invoke(true);
        } catch (Throwable e) {
            Object itemSpigot = null;
            try {
                itemSpigot = itemSpigotMethod.bindTo(im).invoke();
                itemSpigotUnbreakableMethod.bindTo(itemSpigot).invoke(true);
            } catch (Throwable ignored) {
            }
        }
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder setType(Material material) {
        is.setType(material);
        return this;
    }

    public ItemBuilder setLore(String... lore) {
        ItemMeta im = is.getItemMeta();
        im.setLore(new ArrayList<>(Arrays.asList(lore)));
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder setLore(List<String> lore) {
        if (lore == null) {
            return this;
        }
        ItemMeta im = is.getItemMeta();
        im.setLore(lore);
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder removeLoreLine(String line) {
        ItemMeta im = is.getItemMeta();
        List<String> lore = new ArrayList<>(im.getLore());
        if (!lore.contains(line)) {
            return this;
        } else {
            lore.remove(line);
            im.setLore(lore);
            is.setItemMeta(im);
            return this;
        }
    }

    public ItemBuilder removeLoreLine(int index) {
        ItemMeta im = is.getItemMeta();
        List<String> lore = new ArrayList<>(im.getLore());
        if (index >= 0 && index < lore.size()) {
            lore.remove(index);
            im.setLore(lore);
            is.setItemMeta(im);
            return this;
        } else {
            return this;
        }
    }

    public ItemBuilder addLoreLine(String line) {
        ItemMeta im = is.getItemMeta();
        List<String> lore = new ArrayList<>();
        if (im.hasLore()) {
            lore = new ArrayList<>(im.getLore());
        }
        lore.add(ChatColor.translateAlternateColorCodes('&', line));
        im.setLore(lore);
        is.setItemMeta(im);
        return this;
    }

    public ItemBuilder addLoreLine(String line, int pos) {
        ItemMeta im = is.getItemMeta();
        List<String> lore = new ArrayList<>(im.getLore());
        lore.add(pos, line);
        im.setLore(lore);
        is.setItemMeta(im);
        return this;
    }

    public int getLores() {
        ItemMeta im = is.getItemMeta();
        return im != null && im.hasLore() ? im.getLore().size() : 0;
    }

    public ItemBuilder addLoreLine(String... lines) {
        for (String line : lines) {
            addLoreLine(line);
        }
        return this;
    }

    public ItemBuilder setDyeColor(DyeColor color) {
        is.setDurability(color.getDyeData());
        return this;
    }

    @Deprecated
    public ItemBuilder setWoolColor(DyeColor color) {
        if (is.getType().name().equals("WOOL") || is.getType().name().endsWith("_WOOL")) {
            is.setDurability(color.getWoolData());
        }
        return this;
    }

    public ItemBuilder setLeatherArmorColor(Color color) {
        try {
            LeatherArmorMeta im = (LeatherArmorMeta) is.getItemMeta();
            im.setColor(color);
            is.setItemMeta(im);
        } catch (ClassCastException ignored) {
        }
        return this;
    }

    public ItemBuilder setMapView(MapView mapView) {
        setDurability(mapView.getId());
        return this;
    }

    public ItemBuilder addPotion(PotionEffect potionEffect) {
        PotionMeta im = (PotionMeta) is.getItemMeta();
        im.addCustomEffect(potionEffect, true);
        is.setItemMeta(im);
        return this;
    }

    public ItemStack build() {
        return is;
    }

    public ItemBuilder addBannerPattern(Pattern pattern) {
        BannerMeta bannerMeta = (BannerMeta) is.getItemMeta();
        bannerMeta.addPattern(pattern);
        is.setItemMeta(bannerMeta);
        return this;
    }

    public ItemBuilder setBannerBaseColor(DyeColor color) {
        BannerMeta bannerMeta = (BannerMeta) is.getItemMeta();
        bannerMeta.setBaseColor(color);
        is.setItemMeta(bannerMeta);
        return this;
    }

    public ItemBuilder addLore(String... s) {
        return addLoreLine(s);
    }

    public ItemBuilder glow() {
        return addEnchant(Enchantment.DURABILITY, 1).addFlag(ItemFlag.HIDE_ENCHANTS);
    }

    public void line() {
        addLore("§h§u§a§n§m§e§n§g");
    }

    public Material getType() {
        return is.getType();
    }

    public ItemBuilder copy() {
        return new ItemBuilder(is.clone());
    }

    private static MethodHandle itemSpigotMethod;
    private static MethodHandle itemSpigotUnbreakableMethod;
    private static MethodHandle itemUnbreakableMethod;

    private static MethodHandle getPlayerProfileMethod;
    private static MethodHandle setPlayerProfileMethod;

    static {
        try {
            itemSpigotMethod = MethodHandles.lookup().unreflect(ItemStack.class.getDeclaredMethod("spigot"));
        } catch (Exception e) {
            itemSpigotMethod = null;
        }

        try {
            itemSpigotUnbreakableMethod = MethodHandles.lookup().unreflect(ItemMeta.class.getDeclaredMethod("spigot").getReturnType().getDeclaredMethod("setUnbreakable", boolean.class));
        } catch (Exception e) {
            itemSpigotUnbreakableMethod = null;
        }

        try {
            itemUnbreakableMethod = MethodHandles.lookup().unreflect(ItemMeta.class.getDeclaredMethod("setUnbreakable", boolean.class));
        } catch (Exception e) {
            itemUnbreakableMethod = null;
        }

        try {
            getPlayerProfileMethod = MethodHandles.lookup().unreflect(OfflinePlayer.class.getDeclaredMethod("getPlayerProfile"));
        } catch (Exception e) {
            getPlayerProfileMethod = null;
        }

        try {
            Class<?> playerProfileClass = Class.forName("org.bukkit.profile.PlayerProfile");
            setPlayerProfileMethod = MethodHandles.lookup().unreflect(SkullMeta.class.getDeclaredMethod("setOwnerProfile", playerProfileClass));
        } catch (Exception e) {
            setPlayerProfileMethod = null;
        }
    }

}

File: src/main/kotlin/me/huanmeng/opensource/bukkit/util/item/ItemBuilder.kt
package me.huanmeng.opensource.bukkit.util.item

import org.bukkit.ChatColor
import org.bukkit.Color
import org.bukkit.DyeColor
import org.bukkit.Material
import org.bukkit.block.banner.Pattern
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.*
import org.bukkit.map.MapView
import org.bukkit.potion.PotionEffect
import java.util.*
import kotlin.random.Random

/**
 * 2023/3/17<br>
 * Gui<br>
 *
 * @author huanmeng_qwq
 */
open class ItemBuilder : Cloneable {
    private var `is`: ItemStack

    constructor(`is`: ItemStack) {
        this.`is` = `is`.clone()
    }

    constructor(`is`: ItemStack, name: String?) {
        this.`is` = `is`
        this.setName(name)
    }

    constructor(m: Material?, name: String?) {
        `is` = ItemStack(m)
        this.setName(name)
    }

    @JvmOverloads
    constructor(m: Material?, amount: Int = 1) {
        `is` = ItemStack(m, amount)
    }

    constructor(m: Material?, amount: Int, durability: Byte) {
        `is` = ItemStack(m, amount, durability.toShort())
    }

    constructor(type: Material?, amount: Int, durability: Short) {
        `is` = ItemStack(type, amount, durability)
    }

    override fun clone(): ItemBuilder {
        return ItemBuilder(`is`)
    }

    open fun setDurability(dur: Short): ItemBuilder {
        `is`.durability = dur
        return this
    }

    open fun setDurability(dur: Int): ItemBuilder {
        `is`.durability = dur.toShort()
        return this
    }

    open fun setAmount(amount: Int): ItemBuilder {
        `is`.amount = amount
        return this
    }

    open fun setName(str: String?): ItemBuilder {
        var name = str
        val im = `is`.itemMeta
        if (name != null && name.isEmpty()) {
            name = "§c"
        }
        if (name == null) {
            im.displayName = null
        } else {
            im.displayName = ChatColor.translateAlternateColorCodes('&', name)
        }
        `is`.itemMeta = im
        return this
    }

    open fun setNameUnColor(name: String?): ItemBuilder {
        var name = name
        val im = `is`.itemMeta
        if (name != null && name.isEmpty()) {
            name = "§c"
        }
        if (name == null) {
            im.displayName = null
        } else {
            im.displayName = name
        }
        `is`.itemMeta = im
        return this
    }

    val name: String?
        get() {
            val im: ItemMeta? = `is`.itemMeta
            return im?.displayName
        }

    open fun addFlag(vararg flag: ItemFlag?): ItemBuilder {
        val im = `is`.itemMeta
        im.addItemFlags(*flag)
        `is`.itemMeta = im
        return this
    }

    open fun removeFlag(vararg flag: ItemFlag?): ItemBuilder {
        val im = `is`.itemMeta
        im.removeItemFlags(*flag)
        `is`.itemMeta = im
        return this
    }

    open fun addUnsafeEnchantment(ench: Enchantment?, level: Int): ItemBuilder {
        return if (`is`.type == Material.ENCHANTED_BOOK) {
            val im = `is`.itemMeta as EnchantmentStorageMeta
            im.addStoredEnchant(ench, level, true)
            `is`.itemMeta = im
            this
        } else {
            `is`.addUnsafeEnchantment(ench, level)
            this
        }
    }

    open fun addRandomUnsafeEnchantment(level: Int, vararg ench: Enchantment?): ItemBuilder {
        `is`.addUnsafeEnchantment(ench[Random.nextInt(ench.size)], level)
        return this
    }

    open fun clearLores(): ItemBuilder {
        val im = `is`.itemMeta
        im.lore = ArrayList()
        `is`.itemMeta = im
        return this
    }

    open fun removeEnchantment(ench: Enchantment?): ItemBuilder {
        `is`.removeEnchantment(ench)
        return this
    }

    open fun removeEnchantments(): ItemBuilder {
        val im = `is`.itemMeta
        im.enchants.forEach { (enchantment: Enchantment?, integer: Int?) -> im.removeEnchant(enchantment) }
        return this
    }

    open fun setSkullOwner(owner: String?): ItemBuilder {
        try {
            val im = `is`.itemMeta as SkullMeta
            im.owner = owner
            `is`.itemMeta = im
        } catch (_: ClassCastException) {
        }
        return this
    }

    open fun addEnchant(ench: Enchantment?, level: Int): ItemBuilder {
        return if (`is`.type == Material.ENCHANTED_BOOK) {
            val im = `is`.itemMeta as EnchantmentStorageMeta
            im.addStoredEnchant(ench, level, true)
            `is`.itemMeta = im
            this
        } else {
            val im = `is`.itemMeta
            im.addEnchant(ench, level, true)
            `is`.itemMeta = im
            this
        }
    }

    open fun addEnchantments(enchantments: Map<Enchantment?, Int?>?): ItemBuilder {
        `is`.addEnchantments(enchantments)
        return this
    }

    open fun setInfinityDurability(): ItemBuilder {
        `is`.durability = 32767.toShort()
        return this
    }

    open fun setUnbreakable(): ItemBuilder {
        val im = `is`.itemMeta
        im.spigot().isUnbreakable = true
        `is`.itemMeta = im
        return this
    }

    open fun setType(material: Material?): ItemBuilder {
        `is`.type = material
        return this
    }

    open fun setLore(vararg lore: String): ItemBuilder {
        val im = `is`.itemMeta
        im.lore = listOf(elements = lore)
        `is`.itemMeta = im
        return this
    }

    open fun setLore(lore: List<String?>?): ItemBuilder {
        if (lore == null) {
            return this
        }
        val im = `is`.itemMeta
        im.lore = lore
        `is`.itemMeta = im
        return this
    }

    open fun removeLoreLine(line: String?): ItemBuilder {
        val im = `is`.itemMeta
        val lore: MutableList<String?> = ArrayList(im.lore)
        return if (!lore.contains(line)) {
            this
        } else {
            lore.remove(line)
            im.lore = lore
            `is`.itemMeta = im
            this
        }
    }

    open fun removeLoreLine(index: Int): ItemBuilder {
        val im = `is`.itemMeta
        val lore: MutableList<String?> = ArrayList(im.lore)
        return if (index >= 0 && index <= lore.size) {
            lore.removeAt(index)
            im.lore = lore
            `is`.itemMeta = im
            this
        } else {
            this
        }
    }

    open fun addLoreLine(line: String?): ItemBuilder {
        val im = `is`.itemMeta
        var lore: MutableList<String?> = ArrayList()
        if (im.hasLore()) {
            lore = ArrayList(im.lore)
        }
        lore.add(ChatColor.translateAlternateColorCodes('&', line))
        im.lore = lore
        `is`.itemMeta = im
        return this
    }

    open fun addLoreLine(line: String, pos: Int): ItemBuilder {
        val im = `is`.itemMeta
        val lore: MutableList<String> = ArrayList(im.lore)
        lore.add(pos, line)
        im.lore = lore
        `is`.itemMeta = im
        return this
    }

    val lores: Int
        get() {
            val im = `is`.itemMeta
            return if (im.hasLore()) im.lore.size else 0
        }

    open fun addLoreLine(vararg lines: String?): ItemBuilder {
        for (line in lines) {
            this.addLoreLine(line)
        }
        return this
    }

    open fun setDyeColor(color: DyeColor): ItemBuilder {
        `is`.durability = color.dyeData.toShort()
        return this
    }

    @Deprecated("")
    open fun setWoolColor(color: DyeColor): ItemBuilder {
        if (`is`.type == Material.WOOL) {
            `is`.durability = color.woolData.toShort()
        }
        return this
    }

    open fun setLeatherArmorColor(color: Color?): ItemBuilder {
        try {
            val im = `is`.itemMeta as LeatherArmorMeta
            im.color = color
            `is`.itemMeta = im
        } catch (ignored: ClassCastException) {
        }
        return this
    }

    open fun setMapView(mapView: MapView): ItemBuilder {
        this.setDurability(mapView.id)
        return this
    }

    open fun addPotion(potionEffect: PotionEffect?): ItemBuilder {
        val im = `is`.itemMeta as PotionMeta
        im.addCustomEffect(potionEffect, true)
        `is`.itemMeta = im
        return this
    }

    open fun build(): ItemStack {
        return `is`
    }

    open fun addBannerPattern(pattern: Pattern?): ItemBuilder {
        val bannerMeta = `is`.itemMeta as BannerMeta
        bannerMeta.addPattern(pattern)
        `is`.itemMeta = bannerMeta
        return this
    }

    open fun setBannerBaseColor(color: DyeColor?): ItemBuilder {
        val bannerMeta = `is`.itemMeta as BannerMeta
        bannerMeta.baseColor = color
        `is`.itemMeta = bannerMeta
        return this
    }

    fun addLore(vararg s: String): ItemBuilder {
        return this.addLoreLine(lines = s)
    }

    fun glow(): ItemBuilder {
        return addEnchant(Enchantment.DURABILITY, 1).addFlag(ItemFlag.HIDE_ENCHANTS)
    }

    fun line() {
        addLore("§h§u§a§n§m§e§n§g")
    }

    val type: Material
        get() = `is`.type

    fun copy(): ItemBuilder {
        return ItemBuilder(`is`.clone())
    }
}